File: Java/sort/ShellSort.java
import java.util.Scanner;

/*
*Shell Sort is a very efficient sorting algorithm based on insertion sort technique.
*
*Shell Sort is more effective than insertion sort as
*sometimes in Insertion Sorting, we shift a large block to insert an item at the right place 
*as per the sorting order,
*which can be avoided by using the Shell Sort algorithm.
*
*In Shell Sort, the sorting is done at certain intervals/gaps and everytime, this gap is reduced.
*This way, Shell Sort becomes nearly twice as fast as Insertion Sort.
*/

public class ShellSort 
{
	int shellsort(int arr[], int n) 
	{ 
		//here n is the size of the array
		//initially gap = n/2
		for (int gap = n/2; gap > 0; gap = gap/2) 	
		{ 
			for (int i = gap; i < n; i++) 
			{
				//decreasing the gap
				int k = arr[i];
                		int j = i;
                
				while (j >= gap && arr[j - gap] > k) 
				{
				    arr[j] = arr[j - gap];
				    j = j - gap;
				}
				arr[j] = k;	
			} 
		} 
		return 0; 
	} 
    
    	private static Scanner sc = new Scanner(System.in);

	public static void main(String args[]) 
	{ 
		//we store the number of elements in n
		System.out.println("Kindly enter the number of elements: ");
		int n = sc.nextInt();
		
		//declaring an array of n elements
        	int array[] = new int[n];
		
		//taking the input from the user
		for(int i=0;i<n;i++)
		{
			System.out.print("\n" + i + "'th element : ");
			array[i] = sc.nextInt();
		}
		
		System.out.println("\nArray before sorting"); 
		for (int j=0; j<n; j++) 
			System.out.print(array[j] + " "); 
		
		
		//creating an object of shellsort class so that the methods inside shellsort class can be used
		ShellSort object = new ShellSort(); 
		object.shellsort(array,n);
		
		System.out.println("\nArray after sorting"); 
		for (int k=0; k<n; k++) 
			System.out.print(array[k] + " "); 
		
	} 
}
/*

Output:

Kindly enter the number of elements: 5

0'th element : 1
1'th element : 2
2'th element : 5
3'th element : 3
4'th element : 0
Array before sorting
1 2 5 3 0 
Array after sorting
0 1 2 3 5 
*/ 

File: Dart/sort/Shell_Sort.dart
/* 
Shell Sort in Dart

Shell sort is an algorithm that first sorts the elements far apart from 
each other and successively reduces the interval between the elements to 
be sorted. 
*/

import 'dart:io';

void ShellSort(var arr, int n) {
  // Setting the intervals as n/2, n/4, n/8, ...
  for (int interval = (n / 2).floor();
      interval > 0;
      interval = (interval / 2).floor()) {
    // traversing through the right side of the interval
    for (int i = interval; i < n; i++) {
      int temp = arr[i];

      int j;

      // Sorting the temp element by comparing it with left side of the interval
      for (j = i; j >= interval && arr[j - interval] > temp; j -= interval) {
        arr[j] = arr[j - interval];
      }
      arr[j] = temp;
    }
  }
}

void main() {
  var array = [];

  var n, ele;
  print('Enter the number of Elements: ');
  n = stdin.readLineSync();
  n = int.parse(n);

  for (int i = 1; i <= n; i++) {
    print('Enter Element $i: ');
    ele = stdin.readLineSync();
    ele = int.parse(ele);
    array.add(ele);
  }

  ShellSort(array, n);
  print(array);
}

/*
Time Complexity:
1. Worst Case Complexity: O(n^2)
2. Best Case Complexity: O(n*log n)
3. Average Case Complexity: O(n*log n)

Space Complexity: O(1) 

Sample Input/Output:
Input :
Enter the number of Elements: 
5
Enter Element 1: 
50
Enter Element 2: 
30
Enter Element 3: 
10
Enter Element 4: 
40
Enter Element 5: 
20
Output:
[10, 20, 30, 40, 50]

*/
